core/runtime/register_trt_op.cpp

```diff
@@ -10,80 +10,88 @@ namespace trtorch {
     if (curr_device.device_name != conf_device.device_name) {
       LOG_WARNING(
-          "TRTEngine compiled for " << conf_device.device_name << " but current CUDA device is "
-                                    << curr_device.device_name << ". Switching the device context");
+          "Program compiled for " << conf_device.device_name << " but current CUDA device is " << curr_device
+                                  << ". Switching the device context");
       return true;
     }
   }
 
-  if (curr_device.id != conf_device.id) {
-    LOG_WARNING(
-        "Configured Device ID: " << conf_device.id << " is different that current device ID: " << curr_device.id
-                                 << ". Switching context");
-    return true;
-  }
+  // REVIEW: This shouldnt be a reason to switch GPU
+  // if (curr_device.id != conf_device.id) {
+  //   LOG_WARNING(
+  //       "Configured Device ID: " << conf_device.id << " is different that current device ID: " << curr_device.id
+  //                                << ". Switching context");
+  //   return true;
+  // }
 
   return false;
 }
 
-int select_cuda_device(const CudaDevice& conf_device) {
-  int device_id = 0;
-  auto dla_supported = get_dla_supported_SM();
+CudaDevice select_cuda_device(const CudaDevice& conf_device) {
+  int64_t device_id = -1;
+  auto dla_supported = get_dla_supported_SMs();
+
+  auto device_list = get_available_device_list().get_devices();
 
-  auto device_list = cuda_device_list.instance().get_devices();
+  CudaDevice new_target_device;
 
   for (auto device : device_list) {
-    auto compute_cap = std::to_string(device.second.major) + "." + std::to_string(device.second.minor);
+    auto compute_cap = device.second.getSMCapability();
     // In case of DLA select the DLA supported device ID
     if (conf_device.device_type == nvinfer1::DeviceType::kDLA) {
       if (dla_supported.find(compute_cap) != dla_supported.end() &&
           dla_supported[compute_cap] == device.second.device_name) {
         device_id = device.second.id;
+        new_target_device = CudaDevice(device_id, nvinfer1::DeviceType::kDLA);
         break;
       }
     } else if (conf_device.device_type == nvinfer1::DeviceType::kGPU) {
-      auto conf_sm = std::to_string(conf_device.major) + "." + std::to_string(conf_device.minor);
+      auto conf_sm = conf_device.getSMCapability();
       if (compute_cap == conf_sm && device.second.device_name == conf_device.device_name) {
         device_id = device.second.id;
+        new_target_device = CudaDevice(device_id, nvinfer1::DeviceType::kGPU);
         break;
       }
     } else {
-      LOG_ERROR("Unkown device type detected from the compiled engine");
+      TRTORCH_THROW_ERROR("Unknown target device type detected from the compiled program (runtime.select_cuda_device)");
       break;
     }
   }
-  return device_id;
+
+  // REVIEW: THIS DOES NOT LIST DLA PROBABLY, WHICH WE SHOULD
+  TRTORCH_CHECK(
```

Comment:
We should figure out a way to add DLA to the index or have a DLA index or something, I think the debugability for GPUs is good now but for DLA it kind of sucks

Reply:
From a TRT perspective, DLA is configurable using the GPU id. Even if we add a mechanism for DLA deviceType. Internally we would still be validating against Xavier's GPU id.

It makes sense to keep track of DLA Core instead which is already taken care of. (Use Case: Hybrid device configuration for inference)